pkg/sql/executor_ir.go

```diff
@@ -187,242 +147,11 @@ func runSingleSQLIR(wr *PipeWriter, sqlIR ir.SQLStatement, db *DB, modelDir stri
-		return err
-	}
-	defer os.RemoveAll(cwd)
-
-	if submitter() != SubmitterDefault && submitter() != SubmitterPAI {
-		return runThirdPartySubmitterPred(wr, predIR.OriginalSQL, db, cwd, session)
-	}
-
-	// ------------------- run pred IR -----------------------
-	var program bytes.Buffer
-	if submitter() == SubmitterPAI {
-		code, err := pai.Predict(predIR, predIR.TrainIR.Into, cwd)
-		if err != nil {
-			return err
-		}
-		err = createPredictionTableFromIR(predIR, db, session)
-		if err != nil {
-			return err
-		}
-
-		program.WriteString(code)
-	} else {
-		if err := recoverModelDir(db, cwd, modelDir, predIR.TrainIR.Into); err != nil {
-			return err
-		}
-		if isXGBoostModel(predIR.TrainIR.Estimator) {
-			code, err := xgboost.Pred(predIR, session)
-			if err != nil {
-				return err
-			}
-			err = createPredictionTableFromIR(predIR, db, session)
-			if err != nil {
-				return err
-			}
-			program.WriteString(code)
-		} else {
-			err = createPredictionTableFromIR(predIR, db, session)
-			if err != nil {
-				return err
-			}
-			code, err := tensorflow.Pred(predIR, session)
-			if err != nil {
-				return err
-			}
-			program.WriteString(code)
-		}
-	}
-
-	var buf bytes.Buffer
-	buf.WriteString(fmt.Sprintf("\n==========Program======\n%s\n=======Program Output===========\n", program.String()))
-
-	cw := &logChanWriter{wr: wr}
-	w := io.MultiWriter(cw, &buf)
-	defer cw.Close()
-	cmd := sqlflowCmd(cwd, db.driverName)
-	cmd.Env = append(os.Environ())
-	cmd.Stdin = &program
-	cmd.Stdout = w
-	cmd.Stderr = w
-	if e := cmd.Run(); e != nil {
-		return fmt.Errorf("predict failed: %v\n %s", e, buf.String())
-	}
-	return nil
-}
-
-func runAnalyzeIR(analyzeIR *ir.AnalyzeClause, wr *PipeWriter, db *DB, modelDir string, session *pb.Session) error {
-	// cwd is used to load the saved model for prediction.
-	cwd, err := ioutil.TempDir("/tmp", "sqlflow")
-	if err != nil {
-		return err
-	}
-	defer os.RemoveAll(cwd)
-
-	// load the model for analyze
-	if err := recoverModelDir(db, cwd, modelDir, analyzeIR.TrainIR.Into); err != nil {
-		return err
-	}
-
-	cmd := exec.Command("python", "-u")
-	cmd.Dir = cwd
-
-	if !strings.HasPrefix(strings.ToUpper(analyzeIR.TrainIR.Estimator), `XGBOOST.`) {
-		return fmt.Errorf("unsupported model %s", analyzeIR.TrainIR.Estimator)
-	}
-	code, err := xgboost.Analyze(analyzeIR)
-	if err != nil {
-		return err
-	}
-	var program bytes.Buffer
-	program.WriteString(code)
-	cmd.Stdin = &program
-	if _, err := cmd.CombinedOutput(); err != nil {
-		return err
-	}
-
-	imgFile, err := os.Open(path.Join(cwd, "summary.png"))
-	if err != nil {
-		return err
-	}
-	defer imgFile.Close()
-
-	imgBytes, err := ioutil.ReadAll(imgFile)
-	if err != nil {
-		return err
-	}
-	imgBase64Str := base64.StdEncoding.EncodeToString(imgBytes)
-	img2html := fmt.Sprintf("<div align='center'><img src='data:image/png;base64,%s' /></div>", imgBase64Str)
-	wr.Write(img2html)
-	return nil
+	defer submitter().Teardown()
+	return sqlIR.Execute(submitter())
```

Comment:
We can execute a program, execute(interpret) the intermediate representation, but not execute a submitter.

Reply:
> We can execute a program, execute(interpret) the intermediate representation, but not execute a submitter.

Execute is an intransitive and transitive verb.
Transitive verb example:
> A system that executes a program is called an interpreter of the program. ✅

Intransitive verb example:
> Very few programs execute on a bare machine.✅

Both the examples are quoted from https://en.wikipedia.org/wiki/Execution_(computing)
So we can say _The intermediate representation executes **on** a submitter_.

Reply:
> So we can say The intermediate representation executes on a submitter.

@typhoonzero @shendiaomo How about changing `sqlIR.Execute(submitter())` to `sqlIR.ExecuteOn(submitter())`.